futures-mutex/src/lib.rs

```diff
@@ -112,11 +60,18 @@ impl<T> Mutex<T> {
     ///
     /// This function will panic if called outside the context of a future's task.
     pub fn poll_lock(&self) -> Async<MutexGuard<T>> {
-        if self.inner.locked.compare_and_swap(false, true, Ordering::SeqCst) {
-            self.inner.wait_queue.push(current());
-            Async::NotReady
-        } else {
-            Async::Ready(MutexGuard{ inner: self })
+        let mut ext_lock = self.inner.wait_queue.lock();
+        match self.inner.data.try_lock() {
+            Some(guard) => {
+                Async::Ready(MutexGuard {
+                    inner: &self.inner,
+                    guard: Some(guard),
+                })
+            },
+            None => {
+                ext_lock.push(current());
```

Comment:
Can we `use futures::task` so this is `task::current()`? I know that you don't like using module names but in this case it really increases readability. At the moment a casual reader would think, "current what?"